src/MongoDB.Driver.Core/Core/Authentication/AuthenticationHelper.cs

```diff
@@ -36,28 +35,92 @@ public static void Authenticate(IConnection connection, ConnectionDescription de
+            {
+                Authenticator = authenticator,
+                // first authenticator already has the SaslAuthMechanisms info from initial isMaster 
+                Description = index != 0 && authenticator is DefaultAuthenticator
+                    ? GetMergedDescription(connection, (DefaultAuthenticator) authenticator, description, cancellationToken)
+                    : description
+            });
+
+            foreach (var a in authenticators)
+            {
+                a.Authenticator.Authenticate(connection, a.Description, cancellationToken);
             }
+                
+        }
+
+        private static async Task<ConnectionDescription> GetMergedDescriptionAsync(
+            IConnection connection,
+            DefaultAuthenticator defaultAuthenticator,
+            ConnectionDescription description,
+            CancellationToken cancellationToken)
+        {
+            var isMasterCommand = new BsonDocument {{"isMaster", 1}}
+                .Merge(defaultAuthenticator.GetSaslSupportedMechanismsQuery());
+            var isMasterProtocol = ConnectionInitializer.MasterCommandToProtocol(isMasterCommand);
+            var isMasterResult = await
+                ConnectionInitializer.GetIsMasterResultAsync(connection, isMasterProtocol, cancellationToken)
+                    .ConfigureAwait(false);
+            return new ConnectionDescription(
+                connection.ConnectionId,
+                description.IsMasterResult.MergeSaslSupportedMechanisms(isMasterResult.Wrapped),
+                description.BuildInfoResult);
         }
 
-        public static async Task AuthenticateAsync(IConnection connection, ConnectionDescription description, CancellationToken cancellationToken)
+        private static ConnectionDescription GetMergedDescription(
+            IConnection connection,
+            DefaultAuthenticator defaultAuthenticator,
+            ConnectionDescription description,
+            CancellationToken cancellationToken)
         {
+            var isMasterCommand = new BsonDocument {{"isMaster", 1}}
+                .Merge(defaultAuthenticator.GetSaslSupportedMechanismsQuery());
+            var isMasterProtocol = ConnectionInitializer.MasterCommandToProtocol(isMasterCommand);
+            var isMasterResult =
+                ConnectionInitializer.GetIsMasterResult(connection, isMasterProtocol, cancellationToken);
+            return new ConnectionDescription(
+                connection.ConnectionId,
+                description.IsMasterResult.MergeSaslSupportedMechanisms(isMasterResult.Wrapped),
+                description.BuildInfoResult);
+        }
+
+        public static async Task AuthenticateAsync(
+            IConnection connection,
+            ConnectionDescription description,
+            CancellationToken cancellationToken)
+        {
+            // is defensive programming needed for an internal helper class?
```

Comment:
Forgot to remove this comment before starting the review, so I'll remove it with a commit post-review, but this was just a question that came to mind.